doc/security/implementation_guidelines/hardware/index.md

```diff
@@ -0,0 +1,409 @@
+    determine the resulting worst-case scenario. This is a very conservative
+    approach and might lead to over-pessimism, but serves to highlight potential
+    issues. Then, to ease implementation burden, assume the adversary can glitch
+    to all-1's or all-0's (since these are considered "easier" to reach), and
+    that reset can be asserted semi-arbitrarily.
+2.  Use parity/ECC on memories and data paths (note here that ECC is not true
+    integrity, have to use hash to prevent forgery, see [1]). For memories, ECC
+    is helpful to protect instruction streams or values that can cause
+    "branching control flows" that redirect execution flow. Parity is
+    potentially helpful if detection of corruption is adequate (though
+    double-glitch fault injection can fool parity, so Hsiao or other
+    detect-2-error codes can be used, even without correction circuitry
+    implemented). When committing to an irrevocable action (e.g. burning into
+    OTP, unlocking part of the device/increasing permissions), ECC is probably
+    more appropriate.
+    1.  When selecting a specific ECC implementation, the error detection
+        properties are likely more important than error correction (assuming
+        memory lifetime retention/wear are not considered). For a good example
+        of how to consider the effectiveness of error correction, see
+        [this PR comment](https://github.com/lowRISC/opentitan/pull/3899#issuecomment-716799810).
+3.  State machines:
+    1.  Have a minimum Hamming distance for state machine transitions, to make
+        single bit faults non-effective
+    2.  Use a
+        [sparsely populated state encoding](https://github.com/lowRISC/opentitan/blob/master/util/design/sparse-fsm-encode.py),
+        with all others marked invalid - see 11.1 about optimization concerns
+        when doing this though
+    3.  All states could have the same Hamming weight, then can constantly check
+        for this property (or use ECC-type coding on state variable and check
+        this)
+    4.  If waiting for a counter to expire to transition to the next state,
+        better if the terminal count that causes the transition is not
+        all-0/all-1. One could use an LFSR instead of a binary counter, but
+        debugging this can be a bit painful then
+4.  Maintain value-and-its-complement throughout datapath (sometimes called
+    "dual rail" logic), especially if unlocking/enabling something sensitive,
+    and continually check for validity/consistency of representation
+5.  Incorporate temporal randomness where possible (example: add delay cycles
+    based on LFSR around sensitive operations)
+6.  Run-it-twice and compare results for sensitive calculations
+7.  Redundancy - keep/store multiple copies of sensitive checks/data
+8.  For maximum sensitivity, compare combinational and sequential paths with
+    hair-trigger/one-shot latch of miscompare
```

Comment:
Do we have an example in OpenTitan where we're considering doing this? It sounds "rather physical" to me, and therefore scary! So it would be nice to be able to point at an example.